File: backend/unified/unified_test.go
package unified

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestBuildStoredPayload(t *testing.T) {
	messages := []map[string]any{
		{"role": "system", "content": "be brief"},
		{"role": "user", "content": "what is 2+2"},
		{"role": "assistant", "content": "4", "tool_calls": []any{
			map[string]any{"id": "c1", "type": "function", "function": map[string]any{"name": "calc", "arguments": "{}"}},
		}},
		{"role": "tool", "tool_call_id": "c1", "content": "ok"},
		{"role": "assistant", "content": "The answer is 4", "reasoning": "adding"},
	}
	prompt, response, tokens := BuildStoredPayload("gpt", messages, nil)

	var p struct {
		Schema   string           `json:"schema"`
		Model    string           `json:"model"`
		Messages []map[string]any `json:"messages"`
	}
	if err := json.Unmarshal([]byte(prompt), &p); err != nil {
		t.Fatalf("prompt not valid json: %v", err)
	}
	// rows are stamped with the unified trace schema (matches frontend)
	if p.Schema != Schema {
		t.Fatalf("expected schema %q, got %q", Schema, p.Schema)
	}
	if p.Model != "gpt" || len(p.Messages) != 4 {
		t.Fatalf("expected 4 history messages, got %d (%s)", len(p.Messages), prompt)
	}
	// mid-conversation tool_calls preserved, with call id
	tcs, ok := p.Messages[2]["tool_calls"].([]any)
	if !ok || len(tcs) != 1 {
		t.Fatalf("expected tool_calls preserved on history assistant: %s", prompt)
	}
	if tc, _ := tcs[0].(map[string]any); tc["id"] != "c1" {
		t.Fatalf("expected tool call id preserved: %s", prompt)
	}
	// tool result carries linkage
	if p.Messages[3]["tool_call_id"] != "c1" {
		t.Fatalf("expected tool_call_id on tool msg: %s", prompt)
	}
	// final assistant becomes response with reasoning_content
	if !strings.Contains(response, "The answer is 4") || !strings.Contains(response, `"reasoning_content":"adding"`) {
		t.Fatalf("unexpected response: %s", response)
	}
	if tokens <= 0 {
		t.Fatalf("expected positive token estimate")
	}
}

func TestBuildStoredPayloadUnifiedFields(t *testing.T) {
	messages := []map[string]any{
		{"role": "user", "content": "look at this", "image": "data:image/png;base64,AAA"},
		{"role": "assistant", "content": "", "tool_calls": []any{
			// non-string arguments get stringified, missing name defaults
			map[string]any{"id": "c9", "function": map[string]any{"name": "calc", "arguments": map[string]any{"x": 1}}},
		}},
		{"role": "user", "content": ""}, // payload-less: dropped
		{"role": "assistant", "content": "done", "reasoning_content": "already unified"},
	}
	prompt, response, _ := BuildStoredPayload("", messages, nil)

	var p struct {
		Model    string           `json:"model"`
		Messages []map[string]any `json:"messages"`
	}
	if err := json.Unmarshal([]byte(prompt), &p); err != nil {
		t.Fatalf("prompt not valid json: %v", err)
	}
	if p.Model != "trace" {
		t.Fatalf("expected model fallback 'trace', got %q", p.Model)
	}
	if len(p.Messages) != 2 {
		t.Fatalf("expected empty message dropped (2 history messages), got %d: %s", len(p.Messages), prompt)
	}
	// image moves into the unified images array; content stays a plain string
	imgs, _ := p.Messages[0]["images"].([]any)
	if len(imgs) != 1 || imgs[0] != "data:image/png;base64,AAA" {
		t.Fatalf("expected image in images field: %s", prompt)
	}
	if _, isStr := p.Messages[0]["content"].(string); !isStr {
		t.Fatalf("expected plain string content: %s", prompt)
	}
	// tool call arguments normalized to a JSON string
	tcs, _ := p.Messages[1]["tool_calls"].([]any)
	tc, _ := tcs[0].(map[string]any)
	fn, _ := tc["function"].(map[string]any)
	if fn["arguments"] != `{"x":1}` {
		t.Fatalf("expected stringified arguments: %s", prompt)
	}
	if !strings.Contains(response, `"reasoning_content":"already unified"`) {
		t.Fatalf("expected reasoning_content alias in response: %s", response)
	}
}

func TestBuildStoredPayloadSourceEnvelope(t *testing.T) {
	messages := []map[string]any{
		{"role": "user", "content": "hi"},
		{"role": "assistant", "content": "hello"},
	}
	raw := `{"sessionId":"s1","message":{"role":"user","content":"hi"}}` + "\n" + `{"message":{"role":"assistant","content":"hello"}}`
	src := SanitizeSource(&SourceEnvelope{Format: "claude-code", Kind: "jsonl", Name: "session.jsonl", Text: raw})

	prompt, _, withTokens := BuildStoredPayload("m", messages, src)
	_, _, withoutTokens := BuildStoredPayload("m", messages, nil)

	var p struct {
		Source *SourceEnvelope `json:"source"`
	}
	if err := json.Unmarshal([]byte(prompt), &p); err != nil {
		t.Fatalf("prompt not valid json: %v", err)
	}
	// verbatim, byte-for-byte
	if p.Source == nil || p.Source.Text != raw || p.Source.Format != "claude-code" || p.Source.Name != "session.jsonl" {
		t.Fatalf("expected verbatim source envelope, got %+v", p.Source)
	}
	// the source copy must not inflate contribution credit
	if withTokens != withoutTokens {
		t.Fatalf("source envelope changed token estimate: %d != %d", withTokens, withoutTokens)
	}
	// invalid envelopes are dropped
	if SanitizeSource(&SourceEnvelope{Text: ""}) != nil || SanitizeSource(nil) != nil {
		t.Fatalf("expected empty envelope to sanitize to nil")
	}
	if s := SanitizeSource(&SourceEnvelope{Text: "x", Kind: "weird"}); s == nil || s.Kind != "jsonl" || s.Format != "trace" {
		t.Fatalf("expected kind/format fallbacks, got %+v", s)
	}
}

func TestNormalizeMessageBlocks(t *testing.T) {
	// Anthropic-style block content: thinking + text + tool_use, then a user
	// message carrying a tool_result block that must split into a tool message,
	// plus a multimodal image message.
	out := NormalizeMessages([]map[string]any{
		{"role": "assistant", "content": []any{
			map[string]any{"type": "thinking", "thinking": "plan"},
			map[string]any{"type": "text", "text": "running"},
			map[string]any{"type": "tool_use", "id": "t1", "name": "bash", "input": map[string]any{"cmd": "ls"}},
		}},
		{"role": "user", "content": []any{
			map[string]any{"type": "tool_result", "tool_use_id": "t1", "content": "a.txt"},
		}},
		{"role": "user", "content": []any{
			map[string]any{"type": "text", "text": "see this"},
			map[string]any{"type": "image_url", "image_url": map[string]any{"url": "data:x"}},
		}},
	})
	if len(out) != 3 {
		t.Fatalf("expected 3 normalized messages, got %d: %+v", len(out), out)
	}
	a := out[0]
	if a["content"] != "running" || a["reasoning_content"] != "plan" {
		t.Fatalf("unexpected assistant normalization: %+v", a)
	}
	tcs, _ := a["tool_calls"].([]map[string]any)
	if len(tcs) != 1 || tcs[0]["id"] != "t1" {
		t.Fatalf("expected tool call with id from block: %+v", a)
	}
	tm := out[1]
	if tm["role"] != "tool" || tm["content"] != "a.txt" || tm["tool_call_id"] != "t1" {
		t.Fatalf("expected split tool message: %+v", tm)
	}
	um := out[2]
	imgs, _ := um["images"].([]any)
	if um["content"] != "see this" || len(imgs) != 1 || imgs[0] != "data:x" {
		t.Fatalf("expected flattened multimodal user message: %+v", um)
	}
}

func TestPromptJSON(t *testing.T) {
	rawReq := `{"model":"gpt","messages":[{"role":"user","content":"hi"}],"stream":true}`
	prompt, tokens := PromptJSON("gpt", []map[string]any{{"role": "user", "content": "hi"}},
		SanitizeSource(&SourceEnvelope{Format: "openai-chat", Kind: "json", Text: rawReq}))

	var p struct {
		Schema   string           `json:"schema"`
		Messages []map[string]any `json:"messages"`
		Source   *SourceEnvelope  `json:"source"`
	}
	if err := json.Unmarshal([]byte(prompt), &p); err != nil {
		t.Fatalf("prompt not valid json: %v", err)
	}
	// no response split: the user turn stays in messages
	if p.Schema != Schema || len(p.Messages) != 1 || p.Messages[0]["content"] != "hi" {
		t.Fatalf("unexpected prompt: %s", prompt)
	}
	if p.Source == nil || p.Source.Text != rawReq {
		t.Fatalf("expected verbatim wire request as source: %s", prompt)
	}
	if tokens <= 0 || tokens > len(prompt)/4 {
		t.Fatalf("expected source-less token estimate, got %d (prompt len %d)", tokens, len(prompt))
	}
}

File: frontend/src/lib/unified.test.ts
import { describe, expect, test } from "bun:test";
import {
  UNIFIED_SCHEMA,
  normalizeMessages,
  buildStoredPayload,
  toStored,
  storedToMessages,
  toChatCompletions,
  sourceOf,
  sourceFileName,
  type SourceEnvelope,
  type UnifiedMessage,
} from "./unified";

describe("normalizeMessages", () => {
  test("passes through plain chat-completions messages", () => {
    const out = normalizeMessages([
      { role: "system", content: "be brief" },
      { role: "user", content: "hi" },
      { role: "assistant", content: "hello", reasoning_content: "greeting" },
    ]);
    expect(out).toEqual([
      { role: "system", content: "be brief" },
      { role: "user", content: "hi" },
      { role: "assistant", content: "hello", reasoning_content: "greeting" },
    ]);
  });

  test("accepts the pip-proxy `reasoning` alias and tool shapes", () => {
    const out = normalizeMessages([
      {
        role: "assistant",
        content: "",
        reasoning: "let me check",
        tool_calls: [{ id: "c1", type: "function", function: { name: "read", arguments: '{"path":"a"}' } }],
      },
      { role: "tool", content: "file body", tool_call_id: "c1", name: "read" },
    ]);
    expect(out[0]!.reasoning_content).toBe("let me check");
    expect(out[0]!.tool_calls![0]!.id).toBe("c1");
    expect(out[1]).toEqual({ role: "tool", content: "file body", tool_call_id: "c1", name: "read" });
  });

  test("flattens Anthropic-style blocks and splits tool_result into tool messages", () => {
    const out = normalizeMessages([
      { role: "user", content: "run it" },
      {
        role: "assistant",
        content: [
          { type: "thinking", thinking: "plan" },
          { type: "text", text: "running" },
          { type: "tool_use", id: "t1", name: "bash", input: { cmd: "ls" } },
        ],
      },
      { role: "user", content: [{ type: "tool_result", tool_use_id: "t1", content: "a.txt" }] },
      { role: "assistant", content: "done" },
    ]);
    expect(out).toEqual([
      { role: "user", content: "run it" },
      {
        role: "assistant",
        content: "running",
        reasoning_content: "plan",
        tool_calls: [{ id: "t1", type: "function", function: { name: "bash", arguments: '{"cmd":"ls"}' } }],
      },
      { role: "tool", content: "a.txt", tool_call_id: "t1" },
      { role: "assistant", content: "done" },
    ]);
  });

  test("collects images from image_url blocks, image field, and base64 sources", () => {
    const out = normalizeMessages([
      {
        role: "user",
        content: [
          { type: "text", text: "look" },
          { type: "image_url", image_url: { url: "data:image/png;base64,AAA" } },
          { type: "image", source: { type: "base64", media_type: "image/jpeg", data: "BBB" } },
        ],
      },
      { role: "user", content: "and this", image: "https://x/y.png" },
    ]);
    expect(out[0]!.images).toEqual(["data:image/png;base64,AAA", "data:image/jpeg;base64,BBB"]);
    expect(out[1]!.images).toEqual(["https://x/y.png"]);
  });

  test("is deterministic (reproducible): same input, same output bytes", () => {
    const input = [
      { role: "user", content: "q" },
      { role: "assistant", content: "a", tool_calls: [{ function: { name: "t", arguments: { x: 1 } } }] },
    ];
    const a = buildStoredPayload("m", input);
    const b = buildStoredPayload("m", input);
    expect(a).toEqual(b);
  });
});

describe("stored payload", () => {
  test("stamps the schema and splits the final assistant turn", () => {
    const { prompt, response, tokens } = buildStoredPayload("gpt", [
      { role: "user", content: "2+2?" },
      { role: "assistant", content: "4", reasoning: "math" },
    ]);
    const p = JSON.parse(prompt);
    expect(p.schema).toBe(UNIFIED_SCHEMA);
    expect(p.model).toBe("gpt");
    expect(p.messages).toEqual([{ role: "user", content: "2+2?" }]);
    expect(JSON.parse(response)).toEqual({ role: "assistant", content: "4", reasoning_content: "math" });
    expect(tokens).toBeGreaterThan(0);
  });

  test("storedToMessages reassembles the full conversation (back-conversion)", () => {
    const source = [
      { role: "system", content: "sys" },
      { role: "user", content: "q" },
      { role: "assistant", content: "", tool_calls: [{ id: "c1", type: "function", function: { name: "t", arguments: "{}" } }] },
      { role: "tool", content: "out", tool_call_id: "c1" },
      { role: "assistant", content: "a", reasoning_content: "r" },
    ];
    const { prompt, response } = toStored("m", source);
    expect(storedToMessages(prompt, response)).toEqual(source as UnifiedMessage[]);
  });

  test("storedToMessages handles legacy rows (multimodal arrays, JSON strings)", () => {
    const legacyPrompt = JSON.stringify({
      model: "m",
      messages: [
        {
          role: "user",
          content: [
            { type: "text", text: "see image" },
            { type: "image_url", image_url: { url: "data:x" } },
          ],
        },
      ],
    });
    const legacyResponse = JSON.stringify({ role: "assistant", content: "nice", reasoning_content: "" });
    expect(storedToMessages(legacyPrompt, legacyResponse)).toEqual([
      { role: "user", content: "see image", images: ["data:x"] },
      { role: "assistant", content: "nice" },
    ]);
  });
});

describe("toChatCompletions (provider back-conversion)", () => {
  test("re-expands images and keeps tool linkage", () => {
    const unified: UnifiedMessage[] = [
      { role: "user", content: "look", images: ["data:x"] },
      { role: "assistant", content: "", tool_calls: [{ id: "c1", type: "function", function: { name: "t", arguments: "{}" } }] },
      { role: "tool", content: "out", tool_call_id: "c1", name: "t" },
      { role: "assistant", content: "a", reasoning_content: "r" },
    ];
    expect(toChatCompletions(unified)).toEqual([
      {
        role: "user",
        content: [
          { type: "text", text: "look" },
          { type: "image_url", image_url: { url: "data:x" } },
        ],
      },
      { role: "assistant", content: "", tool_calls: [{ id: "c1", type: "function", function: { name: "t", arguments: "{}" } }] },
      { role: "tool", content: "out", tool_call_id: "c1", name: "t" },
      { role: "assistant", content: "a", reasoning_content: "r" },
    ]);
  });
});

describe("source envelope (lossless back-conversion)", () => {
  const jsonl = [
    '{"sessionId":"s1","type":"user","message":{"role":"user","content":"hi"},"cwd":"/tmp","uuid":"u-1"}',
    "not json — kept verbatim anyway",
    '{"sessionId":"s1","type":"assistant","message":{"role":"assistant","content":[{"type":"text","text":"hello"}]},"uuid":"u-2"}',
  ].join("\n");
  const source: SourceEnvelope = { format: "claude-code", kind: "jsonl", name: "session.jsonl", text: jsonl };
  const messages = [
    { role: "user", content: "hi" },
    { role: "assistant", content: "hello" },
  ];

  test("rides along in the stored prompt, byte-for-byte", () => {
    const { prompt } = toStored("m", messages, source);
    expect(prompt.source).toEqual(source);
    // Survives the actual JSON serialization used for storage.
    expect(sourceOf(JSON.stringify(prompt))?.text).toBe(jsonl);
  });

  test("sourceOf reconstructs the envelope from a stored row", () => {
    const { prompt } = toStored("m", messages, source);
    const back = sourceOf(prompt);
    expect(back).toEqual(source);
    expect(sourceFileName(back!)).toBe("session.jsonl");
  });

  test("sourceFileName falls back to format + kind", () => {
    expect(sourceFileName({ format: "codex", kind: "jsonl", text: "x" })).toBe("codex.jsonl");
    expect(sourceFileName({ format: "vscode", kind: "json", text: "x" })).toBe("vscode.json");
  });

  test("does not inflate the token estimate", () => {
    const withSource = toStored("m", messages, { ...source, text: "x".repeat(100_000) });
    const withoutSource = toStored("m", messages);
    expect(withSource.tokens).toBe(withoutSource.tokens);
  });

  test("is omitted when absent or empty", () => {
    expect(toStored("m", messages).prompt.source).toBeUndefined();
    expect(toStored("m", messages, { format: "x", kind: "jsonl", text: "" }).prompt.source).toBeUndefined();
    expect(sourceOf({ model: "m", messages: [] })).toBeNull();
  });

  test("storedToMessages is unaffected by the envelope", () => {
    const { prompt, response } = toStored("m", messages, source);
    expect(storedToMessages(prompt, response)).toEqual(messages as UnifiedMessage[]);
  });
});
